File: libs/gh-toolkit/nexus_gh/release.py
"""The release identity, as the Python half reads it (story #305).

The declaration is a single `VERSION` file at the release root, shared with the TypeScript
executable — the two halves ship as one artifact and cannot be at different versions, so this
module carries no version literal and reads that one file instead. The lookup walks up from this
module's own position, which finds the repository root in a source checkout and the package root
in a distributable without either layout being written down anywhere.

An unresolved declaration is `None`, never a guessed default: an absent version is something a
reader already treats as "written by an unknown toolkit", while a fabricated one is a lie.
"""

from pathlib import Path

#: The file that carries the one declaration, at the release root.
RELEASE_VERSION_FILE = "VERSION"


def resolve_release_version(start_dir: Path) -> str | None:
    """The nearest declaration at or above `start_dir`, or None when there is none."""
    current = Path(start_dir).resolve()
    for candidate_dir in (current, *current.parents):
        candidate = candidate_dir / RELEASE_VERSION_FILE
        if candidate.is_file():
            declared = candidate.read_text(encoding="utf-8").strip()
            return declared or None
    return None


def release_version() -> str | None:
    """The release this toolkit is part of, resolved from where this module itself sits."""
    return resolve_release_version(Path(__file__).resolve().parent)

File: libs/portable-tools/src/release.ts
/**
 * The release identity (story #305). One semantic version covers the TypeScript executable, the
 * Python toolkit and the component payload together, because they ship as one artifact and
 * cannot be at different versions.
 *
 * That version is declared exactly once, in a `VERSION` file at the release root, and both
 * toolkits reach it the same way: walk up from the reader's own file position until a
 * declaration appears. The walk is what makes the one declaration serve both layouts without a
 * build step — in a source checkout it lands on the repository root, and in a distributable it
 * lands on the package root the two toolkits are installed under. Neither half carries a version
 * literal of its own, so there is nothing to keep in step.
 *
 * An unresolved declaration is reported as `null`, never as a guessed or default version: a
 * fabricated version in a writer stamp is worse than an absent one, which a reader already knows
 * how to treat as "written by an unknown toolkit".
 */

import * as fs from "node:fs";
import * as path from "node:path";

/** The file that carries the one declaration, at the release root. */
export const RELEASE_VERSION_FILE = "VERSION";

/** The nearest declaration at or above `startDir`, or null when there is none. */
export function resolveReleaseVersion(startDir: string): string | null {
    let dir: string = path.resolve(startDir);
    for (;;) {
        const candidate: string = path.join(dir, RELEASE_VERSION_FILE);
        if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) {
            const declared: string = fs.readFileSync(candidate, "utf8").trim();
            return declared === "" ? null : declared;
        }
        const parent: string = path.dirname(dir);
        if (parent === dir) {
            return null;
        }
        dir = parent;
    }
}

/** The release this executable is part of, resolved from where this module itself sits. */
export function releaseVersion(): string | null {
    return resolveReleaseVersion(import.meta.dirname);
}
